### ISSUE TITLE
Add convenient way to set config limits

### ISSUE THREAD

**Issue by RAOF** [AltSysrq/proptest#11]
When using the propcheck! macro there doesn't appear to be an easy way to set the values in test_runner::Config.

It would be convenient if these could be inherited from the environment; if, say `PROPCHECK_REQUIRED_SUCCESSES=1024` ran the tests with Config { cases : 1024, ...Config::default() }

(This may be already implemented; I haven't checked. If it *is* implemented then it's not documented 😄 )

**Comment by AltSysrq** [comment]
> When using the propcheck! macro there doesn't appear to be an easy way to set the values in test_runner::Config.

See the bottom of [`proptest!`](https://docs.rs/proptest/0.3.1/proptest/macro.proptest.html). I suppose the top-level docs should mention something about this though since it's not obvious the macro would support extra syntax.

> It would be convenient if these could be inherited from the environment

Very much agreed, I will add this next week. Thanks for the suggestion!

**Comment by Centril** [comment]
Should `PROPTEST_REQUIRED_SUCCESSES` be evaluated at [compile time (when building tests)](https://doc.rust-lang.org/nightly/std/macro.option_env.html) or [runtime (when running tests)](https://doc.rust-lang.org/nightly/std/env/fn.var.html)?

I guess the upside to runtime evaluation is that if you are not satisfied, you can go and change it quickly without rebuilding the test binary. The downside is that there might not be an env variable at runtime (you can fall back to the default amount then).

**Comment by RAOF** [comment]
@Centril Definitely at run-time. If you want to bake different defaults in at build-time, the extra `proptest!` attribute is much more sensible.

You would indeed need to handle the environment variable not being set and using the default `256` in that case.

**Comment by Centril** [comment]
@RAOF So we can simply do:

```rust
const DEFAULT_CASES: u32 = 256;

fn compute_required_cases() -> u32 {
    if let Ok(var) = ::std::env::var("PROPTEST_REQUIRED_CASES") {
        if let Ok(res) = var.parse::<u32>() {
            return res;
        }
    }
    return DEFAULT_CASES;
}

impl Default for Config {
    fn default() -> Config {
        Config {
            cases: compute_required_cases(), // <-- Perhaps use lazy_static!  for this to not recompute - thoughts?
            // ...
        }
    }
}
```

**Comment by RAOF** [comment]
That looks right to me.

I don't *think* you need to bother with `lazy_static!` - reading the environment is cheap (possibly even cheaper than locking and unlocking `lazy_static!`'s mutex ☺).

It *might* be nice to print something if `PROPTEST_REQUIRED_CASES` does *not* successfully parse as a `u32`?

**Comment by Centril** [comment]
Printing sounds nice (when the variable was set (don't print when it's not set) but could not parse).

There's some allocation (for the `String`) and UTF-8 validation going on in [`var`](https://doc.rust-lang.org/nightly/src/std/env.rs.html#205-207). Benchmark or let it be?

**Comment by RAOF** [comment]
How often is a Config created? It's probably not even worth 
benchmarking if it's once per #[test]; it might be if it's once per 
iteration, but that doesn't sound like how it'd work.

**Comment by AltSysrq** [comment]
`lazy_static!` would also be an easy way to ensure the printing only happens once.

I think performance is moot here since a typical proptest usage will call `Config::default()` maybe a few dozen times.

(Though I would like to point out that reading the environment _isn't_ comparatively cheap since it involves walking a linear key-value list and I think requires a mutex on some systems, whereas `lazy_static!` incurs only an atomic load and a branch once it's initialised.)

**Comment by Centril** [comment]
@AltSysrq It also involves an allocation and UTF-8 validation, so it's even worse ^,-
The point about printing once is a good one - I think I'll use it then.

@RAOF It's once per `#[test]`, as seen [here](https://github.com/AltSysrq/proptest/blob/master/src/sugar.rs#L108).

**Comment by AltSysrq** [comment]
This is available in 0.3.2 for all [`Config` fields](https://docs.rs/proptest/0.3.2/proptest/test_runner/struct.Config.html#fields).
